engine/apps/metrics_exporter/metrics_collectors.py

```diff
@@ -196,5 +203,20 @@ def get_buckets_with_sum(self, values):
             sum_value += value
         return buckets_values, sum_value
 
+    def _get_schedule(self):
+        # Initialize the metric object with the updated metric name
+        metrics_schedule = GaugeMetricFamily(
+            METRICS_SCHEDULE, "Description of metric", labels=["schedule", "team", "username"]
+        )
+        # Retrieve on-call schedules data
+        schedules = OnCallSchedule.objects.all()
+        # Process on-call schedule data
+        oncall_users = get_cached_oncall_users_for_multiple_schedules(schedules)
```

Comment:
Revised the code. API calls also use [queryset = OnCallSchedule.objects.all()](https://github.com/grafana/oncall/blob/dev/engine/apps/api/views/schedule.py#L110). There was no caching in the schedulers itself and  `OnCallSchedule.objects.all().query` says that a query is constantly being generated into the database.
Now `/api/v1/schedules/` is in use and does not cause any problems with load or delay in calls. I don’t see any restrictions why the same approach cannot be used now for metrics, since there is no other way.
As for optimization, it seems the focus should be on improving the scheduler itself and implementing caching mechanisms. I looked in there, got scared, and closed it.

Reply:
Hm.. I see, that `queryset` definition there shouldn't be needed and I think it is not being used (note the API class defines and uses the `get_queryset` method, where the query is limited per org).
In any case, the main issue here is that for schedules not in cache, the `get_cached_oncall_users_for_multiple_schedules` may trigger http requests to refresh imported ical files, making this exporter method take too long and then the prometheus scrape would timeout (FWIW, I confirmed this manually running the logic in our stack).
I agree this could work ok in a local setup, but we would need some additional work and optimization before merging and enabling it in cloud (which as I said I would be interested to enable, so I will take a look when I get some time). Makes sense?